Deploy split proxy/auth with helm chart

hugoShaka opened the pull request:
This PR implements big chunks of https://github.com/gravitational/teleport/pull/18274

The following changes are implemented in the PR:
- split responsibilities between two pod sets: pods and proxies
- use the `kubernetes` joinMethod to join proxies
- have a configuration template per mode: aws, gcp, standalone, scratch
- scale proxies by default when they are replicable (cert-manager or user-provided cert through secret, in the future we also might consider pods running behind an ingress as replicable)
- merge user-provided configuration with generated configuration
- fix initContainer logic (the initContainer field was breaking for more than 1 initContainer)

The following changes have been implemented and reviewed in sub-PRs:
- add serviceMonitor for metric collection (see https://github.com/gravitational/teleport/pull/19291)
- add an auth initContainer validating if the configuration is valid (see https://github.com/gravitational/teleport/pull/19333)

The following changes will be implemented later, in subsequent PRs:
- add configuration examples (see https://github.com/gravitational/teleport/pull/19397)
- add a proxy initContainer checking the auth version and blocking rollout (blocked by https://github.com/gravitational/teleport/pull/19277)
- mount short-lived tokens on the proxy using projected volumes when possible (here we are using the default bound tokens with a long expiry)
- documentation updates: (https://github.com/gravitational/teleport/pull/19881)
  - getting started guides
  - reference
  - architecture overview
  - changelog 

The following changes will not be implemented and the RFD will be modified to reflect this change:
- using the operator to create proxy join tokens
- adding a field to deploy CRs alongside the deployment

I am not happy with the state of the tests, but the PR is already too large. I would like to revamp the tests in a subsequent PR.

Before merging I want to run the following tests:
- [x] Follow local lab getting started with old chart and upgrade to split chart
- [x] Follow local lab getting started with split chart
- [x] Follow GKE getting started with old chart and upgrade to split chart
- [x] Follow GKE getting started with split chart
- [x] Follow EKS getting started with old chart and upgrade to split chart
- [x] Follow EKS getting started with split chart (broken, blocked by https://github.com/gravitational/teleport/pull/19188)
- [x] Load-test the chart before and after this PR to estimate the performance impact

Should address:
- https://github.com/gravitational/teleport/issues/16220
- https://github.com/gravitational/teleport/issues/18774
- https://github.com/gravitational/teleport/issues/18569
- https://github.com/gravitational/teleport/issues/18483 (does not provides a happy path but supports it)
- https://github.com/gravitational/teleport/issues/18457
- https://github.com/gravitational/teleport/issues/16871
- https://github.com/gravitational/teleport/issues/16096 (although it should be tested and we might want to add a containerPort to officialise proxy-peering support)
- https://github.com/gravitational/teleport/issues/16095 (does not provides a happy path but support it)
- https://github.com/gravitational/teleport/issues/13260
- https://github.com/gravitational/teleport/issues/9713

hugoShaka committed:
helm: split auth and proxy, templates

hugoShaka committed:
helm: split auth and proxy, lint files

hugoShaka committed:
helm: split auth and proxy, tests

hugoShaka commented on examples/chart/teleport-cluster/templates/auth/_config.common.tpl:
```diff
@@ -0,0 +1,64 @@
+kubernetes_service:
+  enabled: true
+  listen_addr: 0.0.0.0:3026
+  public_addr: "{{ .Release.Name }}-auth.{{ .Release.Namespace }}.svc.cluster.local:3026"
```
I am not happy with the hardcoded cluster domain `cluster.local`. Should this be configurable in value ? Do we want to rely on the ndots behaviour? I'm not sure about how many requests and how many resolutions the proxy performs but this tends to scale terribly.

hugoShaka replied:
I have an idea that will not involve exposing a value but will allow users to override it: I'll put this in a helper template (it's used at 2 or 3 places, so it kinda makes sense to compute it once).

By the way, if you squint hard enough, the chart not only supports `standalone`, `aws`, `gcp` and `scratch` modes, but lets users build their own modes with templating.

If you create your own chart, declare ours as a dependency, and define `teleport-cluster.auth.config.mymode` (or redefine `teleport-cluster.auth.config.common`), you can totally set `chartMode: mymode`. I really don't want to document this because this is not needed by 99% of the users and they will abuse it, but this is definitely a nice trick for Helm power users.

hugoShaka replied:
If I understand correctly the init-container would mount the config, resolve `{{.Release.Name}}-auth`, sed the original config to an `empty-dir` volume and replace a kubernetes_service.public_addr placeholder.

tigrato replied:
Yes, that's my idea

hugoShaka replied:
I think we'll need to build a FQDN it anyway for the proxy initContainer checking if the auth server are updated.

hugoShaka commented on examples/chart/teleport-cluster/templates/auth/_config.common.tpl:
```diff
@@ -0,0 +1,64 @@
+    format:
+      output: {{ .Values.log.format }}
+      extra_fields: {{ .Values.log.extraFields | toJson }}
+kubernetes_service:
```
I let the kubernetes service bundled in the auth for now, but this could be moved out of the deployment.

hugoShaka replied:
As far as I understand, proxies are made to be scaled up; I don't think the Kubernetes service needs to. So I tried to change the rolebindings as few as possible. This is also why it stayed in the auth. If we were to move it, I would prefer to move it to a 3rd deployment, maybe even embed `teleport-kube-agent` or something.

marcoandredinis replied:
>I don't think the Kubernetes service needs to

That was my misconception then.
I thought that Kubernetes Service was meant to provide access to K8s cluster, so the more the merrier (to provide HA)

hugoShaka commented on examples/chart/teleport-cluster/templates/auth/_config.common.tpl:
```diff
@@ -0,0 +1,64 @@
+{{- $authentication := mustMergeOverwrite .Values.authentication (default dict .Values.authenticationSecondFactor) -}}
+{{- $logLevel := (coalesce .Values.logLevel .Values.log.level "INFO") -}}
+version: v3
+teleport:
+  log:
+    severity: {{ $logLevel }}
+    output: {{ .Values.log.output }}
+    format:
+      output: {{ .Values.log.format }}
+      extra_fields: {{ .Values.log.extraFields | toJson }}
```
The teleport block should be at the end of the document as the aws/gcp modes are setting storage, which is in `teleport` and not in `auth_service`

hugoShaka commented:
Upgrading from a past chart caused ~3 min of downtime, mainly caused by:
- auth startuup time: 45s
- proxy connect retry: 30s
- proxy readiness: 30s
- load balacing pool update: 30s

webvictim reviewed (commented):
Looks really good overall. I think the new templating style is going to give people a lot more flexibility.

We might want to update the README a bit to remove the values listing and point people to the [comprehensive chart reference](https://goteleport.com/docs/reference/helm-reference/teleport-cluster/) too... which unfortunately is also going to need overhauling as part of the changes 😢

hugoShaka commented on examples/chart/teleport-cluster/templates/psp.yaml:
```diff
@@ -3,6 +3,7 @@ apiVersion: policy/v1beta1
 kind: PodSecurityPolicy
 metadata:
   name: {{ .Release.Name }}
+  namespace: {{ .Release.Namespace }}
```
Ooof, this is a mistake, PSPs are not namespaced (their roles are). Helm is applying namesapce anyway (except when running `helm template` which caused at least 2 people to report this as a security vuln.

webvictim replied:
OK, in that case I take it all back 😁

hugoShaka replied:
```suggestion
```

tigrato reviewed (commented):
Looks good.

hugoShaka committed:
Apply suggestions from code review

Co-authored-by: Gus Luxton <gus@goteleport.com>

hugoShaka committed:
Allow standalone auth to run with multiple auth pods. Suport AWs lb controller

hugoShaka commented on examples/chart/teleport-cluster/templates/auth/pvc.yaml:
```diff
@@ -0,0 +1,23 @@
+{{- $auth := mustMergeOverwrite .Values .Values.auth -}}
+{{/* $persistence looks like some backward compatibility trick, is this still relevant? */}}
```
I'll do it, but I think for the standalone users, this was not a breaking change prior to this change.

hugoShaka committed:
Treat gus' feedback + avoid auth servicename overflow

marcoandredinis commented on examples/chart/teleport-cluster/tests/__snapshot__/auth_config_test.yaml.snap:
```diff
@@ -0,0 +1,1392 @@
+matches snapshot for auth-webauthn-legacy.yaml:
+  1: |
+    |-
+      Error: 'error converting YAML to JSON: yaml: line 20: mapping values are not allowed
```
Are these errors expected?

hugoShaka replied:
Wow, I thought `helm unittest` would crash if it failed to template and we did not need `make lint-helm` anymore. This is terrible behaviour.

hugoShaka replied:
I fixed the template but did not manage to have a clean fail with neither `lint-helm` or `test-helm`

hugoShaka commented on examples/chart/teleport-cluster/templates/proxy/_config.scratch.tpl:
```diff
@@ -0,0 +1,13 @@
+{{- required "'proxy.teleportConfig' is required in scratch mode" .Values.proxy.teleportConfig }}
+ssh_service:
+  enabled: false
+auth_service:
+  enabled: true
```
Ooooof, nice catch

hugoShaka committed:
Fix webauthn templating, scratch proxy config + marco's feedback

hugoShaka committed:
helm: add PodMonitor support (#19291)

hugoShaka committed:
helm: add job validating configuration on deploy (#19333)

Part of [RFD-0096](https://github.com/gravitational/teleport/pull/18274)

This PR adds helm hooks deploying a test configuration job and running `teleport configure --test` to validate the `teleport.yaml` configuration is sane.

hugoShaka commented:
Load-tested with 5k nodes, no activity:

### Baseline v11 chart
- 2.5 min for full join:
- 5 MiB per tunnel during connecting
- 1.2 MiB per tunnel at rest
- peak at 10 vcpu
- 1.4e-4 vcpu per tunnel at rest

### Split chart for 5k clients
- 2 min for full join:
- proxy:
  - 750 KiB per tunnel during connecting
  - 500 KiB per tunnel at rest
  - peak at 7.2 vcpu
  - 10^-4 vcpu per tunnel at rest
- auth:
  - 1 MiB per registered server during the join
  - 900 KiB per registered server at rest
  - peak at 6.4 vcpu
  - 10^-4 vcpu per registered server at rest
  
 ### Comparison
 
Overall resource consumption is higher in split mode, but per-pod resource consumption is lower. This should not cause sizing issues for users upgrading from v11 chart.

hugoShaka committed:
helm: use a deployment instead of a statefulset for auth + fix podmonitor

hugoShaka committed:
helm: warn that multiplexing requires TLS termination

hugoShaka committed:
helm: fix regression with predeploy job and gcp credentials

hugoShaka commented:
Documentation PR is here: https://github.com/gravitational/teleport/pull/19881

I'll wait until there are no more comments on this one to rebase and send it for review to the docs team.

hugoShaka committed:
fixup! helm: use a deployment instead of a statefulset for auth + fix podmonitor

webvictim commented on examples/chart/teleport-cluster/values.yaml:
```diff
@@ -67,8 +89,12 @@ authentication:
+# TLS multiplexing is not supported when using ACM+NLB for TLS termination.
+#
+# Possible values are 'separate' and 'multiplex'
+proxyListenerMode: "separate"
```
I think more people are terminating TLS in front of Teleport than not when deploying in Kubernetes. The use of ACM is widespread, and the lack of Ingress support is a huge gripe people have with the Helm charts in general. As such I think it's best if we're still explicit about requiring the use of separate ports/listeners until such time as Teleport can work reliably behind TLS termination.

hugoShaka replied:
If/when we'll manage to have Teleport run behind an Ingress we'll be able to multiplex by default on those setups. Based on support requests *a lot* of people are using ACM-based setups and switching to multiplex will break their setups.

webvictim commented on examples/chart/teleport-cluster/templates/auth/_config.scratch.tpl:
```diff
@@ -0,0 +1,13 @@
+{{- end -}}
+
+{{- define "teleport-cluster.auth.config.custom" -}}
+{{ fail "'custom' mode has been removed with chart v12 because of the proxy/auth split breaking change, see http://link" }}
```
Link still needs to be updated to an actual value here, maybe use a bit.ly redirect or something and update it later?

(I wish goteleport.com had its own redirection service that we used in all error messages tbh)

webvictim commented on examples/chart/teleport-cluster/templates/auth/deployment.yaml:
```diff
@@ -0,0 +1,265 @@
+  minReadySeconds: {{ $auth.highAvailability.minReadySeconds }}
+{{- end }}
+  strategy:
+{{- if $replicated }}
```
`type` is set in the `else` case below; does it have to be set here too or does setting `rollingUpdate` remove the need for `type` completely?

hugoShaka replied:
default type is `RollingUpdate`, but it does look clearer when explicitly set on both sides of the `if`

webvictim commented on examples/chart/teleport-cluster/templates/auth/pvc.yaml:
```diff
@@ -0,0 +1,21 @@
+{{- $auth := mustMergeOverwrite .Values .Values.auth -}}
+{{- if $auth.persistence.enabled }}
+  {{/* Disable persistence for aws and gpc modes */}}
```
```suggestion
  {{/* Disable persistence for aws and gcp modes */}}
```

As an aside, I think if we (eventually) end up adding an `azure` chartMode then we should rework all these conditions into some kind of `cloudProvider` "meta" check (via template?) so we don't have to enumerate each individual `chartMode` at multiple different locations in the chart.

hugoShaka replied:
Before we were segregating "stanadalone" and "custom", now it's an implicit category "cloud providers happy path without persistence". I guess it makes sense to at least explicit this, maybe to have a" no persistence mode" list or something.

webvictim commented on examples/chart/teleport-cluster/templates/proxy/_config.common.tpl:
```diff
@@ -0,0 +1,72 @@
+{{- define "teleport-cluster.proxy.config.common" -}}
+{{- $logLevel := (coalesce .Values.logLevel .Values.log.level "INFO") -}}
```
This `coalesce` was a backwards compatibility workaround for when we switched from `logLevel` to `log.level`. I wonder whether it could be removed now to clean things up a bit, given that the use of `logLevel` was deprecated several major versions ago? Not a huge deal if we want to preserve absolute backwards compatibility though.

hugoShaka replied:
I think I'll keep this for now and drop this in v13 to reduce the amount of breaking changes here.

webvictim commented on examples/chart/teleport-cluster/templates/proxy/_config.scratch.tpl:
```diff
@@ -0,0 +1,13 @@
+{{- end -}}
+
+{{- define "teleport-cluster.proxy.config.custom" -}}
+{{ fail "'custom' mode has been removed with chart v12 because of the proxy/auth split breaking change, see http://link" }}
```
Link redirect again

webvictim commented on examples/chart/teleport-cluster/templates/proxy/service.yaml:
```diff
@@ -0,0 +1,66 @@
+{{- $proxy := mustMergeOverwrite .Values .Values.proxy -}}
+{{- $backendProtocol := ternary "ssl" "tcp" (hasKey $proxy.annotations.service "service.beta.kubernetes.io/aws-load-balancer-ssl-cert") -}}
```
I still kinda don't like that we have this "magic" behaviour based solely on the AWS ACM annotation, but we can always tidy this up a bit or make it more explicit in future (maybe once we start being able to support Ingress properly)

hugoShaka replied:
At least now we'll honor the labels you pass instead of forcing ours. This should fix https://github.com/gravitational/teleport/issues/16220

webvictim commented on examples/chart/teleport-cluster/templates/psp.yaml:
```diff
@@ -38,6 +38,7 @@ apiVersion: rbac.authorization.k8s.io/v1
 kind: Role
 metadata:
   name: {{ .Release.Name }}-psp
+  namespace: {{ .Release.Namespace }}
```
Did we decide that PSPs could be namespaced in the end? I thought the idea was that they shouldn't and this wasn't a bug after all.

hugoShaka replied:
It's complicated... PSPs are not namespaced, but their role and role bindings are. This resource is a bit awkward, and I'm happy with it being phased out. This makes me think that we'll need to test against the new built-in security policy thing and see if we need to add labels/annotations to get enrolled in the restricted policy.

webvictim reviewed (approved):
🎉

hugoShaka commented:
Following [discussions](https://github.com/gravitational/teleport/issues/9020), I searched GitHub issues and learned Teleport had several issues with stale proxy/auth nodes in the past. I'll run a couple more tests to observe how the cluster reacts during rollouts and validate that the topology change does not introduce a regression/make things worse.

hugoShaka commented:
### With split auth/proxy

The auth rollout took ~5min for the metrics to go back to normal
![image](https://user-images.githubusercontent.com/16366487/211638513-7e144d8c-7744-43ca-a877-0437ec33b6b2.png)

The proxy rollout happened really fast, and all nodes reconnected real quick. However, CPU usage increased dramatically after the rollout and did not go back to regular levels (for proxy, nodes, and a bit for auth).

According to the logs, the nodes were discovering 3 or 4 proxies for 30 minutes. Once the nodes stopped discovering stale proxies, the CPU usage went back to nominal usage.

![image](https://user-images.githubusercontent.com/16366487/211645267-663e5986-99b4-4a0f-bb59-64b563153b33.png)

hugoShaka committed:
helm: explicit update strategy + add documentation links

zmb3 reviewed (approved):
bot

hugoShaka committed:
Merge branch 'master' into hugo/chart-split-proxy-auth
